### PR TITLE
fix(mcp): declare clio-run args as type=object (ENG-92653)

### PR DESCRIPTION
**Pull request by Alexandr-Kravchuk**
🔗 **Jira:** [ENG-92653](https://creatio.atlassian.net/browse/ENG-92653)

## Summary
- `clio-run` / `clio-run-destructive` `args` parameter was `JsonElement?`, which generates no `"type"` in the JSON Schema
- Claude Code (and other schema-driven MCP clients) require `"type": "object"` to serialize a parameter as a JSON object — without it the args payload is dropped/stringified
- This made the entire hidden-tool surface (~102 tools) unreachable from Claude Code
- Changed parameter type to `Dictionary<string, JsonElement>?` → SDK emits `type: ["object", "null"]`
- Added `DictionaryToElement()` helper to convert before forwarding to executor (executor interface unchanged)

## Schema before/after

**Before:** `"args": { "description": "..." }` — no type declaration

**After:**
```json
"args": {
  "description": "JSON arguments object the target tool expects",
  "type": ["object", "null"],
  "default": null
}
```

## Tests added (4)
- `ClioRunTool_ShouldDeclareArgsAsObjectType_InEmittedInputSchema` — schema regression guard (builds tool via SDK, asserts emitted schema)
- `ClioRunDestructiveTool_ShouldDeclareArgsAsObjectType_InEmittedInputSchema` — same for destructive alias
- `ClioRunTool_ShouldConvertDictionaryArgs_WhenArgsAreProvided` — verifies Dictionary→JsonElement conversion
- `ClioRunTool_ShouldPassNullArgs_WhenDictionaryIsNull` — null path

## Test plan
- [x] `dotnet test --filter "Category=Unit & Module=McpServer"` — 1692 passed
- [x] `dotnet test --filter "Category=Unit & Module=Command"` — 1837 passed
- [ ] Live Claude Code test: repeat ticket repro steps (flat form, wrapped form, control) against fresh build

## Why this was missed in PR #743
Unit tests called `RunAsync(...)` directly with pre-constructed `JsonElement` objects, bypassing the SDK parameter binding and emitted schema entirely. The new schema guard tests build the tool through the SDK and assert the real emitted schema — a type regression will now fail the build.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

[ENG-92653]: https://creatio.atlassian.net/browse/ENG-92653?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: clio/Command/McpServer/Tools/ClioRunTool.cs [hunk 1]
```diff
@@ -468,9 +468,12 @@ public sealed class ClioRunTool(IClioRunExecutor executor) {
 	public ValueTask<CallToolResult> Run(
 		RequestContext<CallToolRequestParams> context,
 		[Description("clio MCP tool name (kebab-case), e.g. \"execute-esq\"")] string? command = null,
-		[Description("JSON arguments object the target tool expects")] JsonElement? args = null,
+		[Description("JSON arguments object the target tool expects")] Dictionary<string, JsonElement>? args = null,
```
Comment (tetiana-moshon) [inline clio/Command/McpServer/Tools/ClioRunTool.cs:471]: **Major — mandatory `clio.mcp.e2e` coverage missing for the changed tool (repo policy).**

`clio/Command/McpServer/AGENTS.md` makes this non-negotiable: *"Every new or changed MCP tool must ship with updated `clio.mcp.e2e` coverage… MCP implementation work is incomplete until the real `clio mcp-server` path is exercised end to end."* Root `AGENTS.md` repeats it ("mandatory even when the user does not mention E2E coverage explicitly").

This PR changes the `args` binding + emitted schema of `clio-run`/`clio-run-destructive`, but adds only unit tests. That is the exact blind spot this bug slipped through in PR #743 — the PR body notes the unit tests "bypassed the SDK parameter binding and emitted schema entirely." The new `McpServerTool.Create(...)` schema guard is a real improvement, but it still builds the tool in-process, not over the wire; there is no `clio.mcp.e2e` test that dispatches through the real `mcp-server` for `clio-run`, and the "Live Claude Code test" checkbox in the test plan is unchecked.

Fix direction: add a `ClioRunToolE2ETests` in `clio.mcp.e2e` that dispatches a flat-form and a wrapped-form `clio-run` call through the real server path and asserts the `args` object reaches the target tool (and, if the harness exposes it, that the emitted schema declares `args` type=object).

**Review by tetiana-moshon — CHANGES_REQUESTED** [review]
## Creatio Code Review — `request changes`

**Decision:** request changes (one high-confidence Major). The production change is small, correct, and well-reasoned; the blocker is a mandated process gap, not a code defect.

### Findings

**1. Major — missing mandatory `clio.mcp.e2e` coverage for the changed MCP tools.**
`clio/Command/McpServer/AGENTS.md` and root `AGENTS.md` make E2E coverage mandatory for every changed MCP tool ("incomplete until the real `clio mcp-server` path is exercised end to end"). This PR changes the `args` binding and emitted schema of `clio-run`/`clio-run-destructive` but adds only unit tests. This is precisely the blind spot that let the bug ship in PR #743 (the PR body itself notes the old tests bypassed SDK binding and the emitted schema). The new `McpServerTool.Create(...)` schema guard is a genuine improvement but still runs in-process, not over the protocol. Fix: add a `clio.mcp.e2e` test dispatching flat-form and wrapped-form `clio-run` calls through the real server. (inline comment on `ClioRunTool.cs:471`)

**2. Minor — friendlier malformed-args error is now unreachable via the tool path.**
With `args` typed as `Dictionary<string,JsonElement>?`, a non-object `args` (array/scalar/stringified JSON) now fails at SDK deserialization instead of reaching the executor's structured `ArgumentException` ("'args' … must be a JSON object", `ClioRunTool.cs:397-400`) and `RecoverWrappedCall`'s tolerant path. Compliant clients are unaffected (that is the point of the fix), but non-compliant callers get a lower-quality error. Consider noting this in the PR/commit or accepting explicitly.

### Verification performed
Traced the full dispatch path against source: both the flat shape `{"command":"X","args":{...}}` and the wrapped shape `{"args":{"command":"X","args":{...}}}` round-trip losslessly through `DictionaryToElement` → `RecoverWrappedCall`/`BuildChildParams`; the null path is preserved; the self/cross-dispatch recursion guard is untouched. `SerializeToElement` on a `Dictionary<string,JsonElement>` is lossless (values are verbatim `JsonElement`), so the missing MCP serializer options are immaterial. No security, performance, or CLIO-analyzer concerns in the diff. Unit tests are well-formed (AAA, `because`, `[Description]`, SDK-built schema guard for both tools).

### Requirement coverage (ENG-92653)
- `args` declared `type` includes `object` in emitted schema — **covered** (schema guard for both tools)
- Dictionary→JsonElement conversion (non-null + null) — **covered**
- Both `clio-run` and `clio-run-destructive` — **covered** (destructive conversion shares the tested `DictionaryToElement` helper)
- Real-client / real-protocol validation ("Live Claude Code test") — **missing** (unchecked in test plan; no `clio.mcp.e2e` coverage — see Finding 1)

### Review summary
Correct, minimal, well-documented fix for a real schema-serialization bug, with a solid unit-level regression guard. Requesting changes solely because the repository makes `clio.mcp.e2e` coverage mandatory for changed MCP tools and this is exactly the layer that would have caught the original defect — unit tests here are necessary but, per policy, insufficient. Once an E2E test exercises `clio-run` over the real server path, this is ready to merge. Confidence: high; the diff is small and was reviewed exhaustively against source (focused single-pass review appropriate to a 7-line production change, not a five-subagent fan-out).

File: clio/Command/McpServer/Tools/ClioRunTool.cs [hunk 2]
```diff
@@ -468,9 +468,12 @@ public sealed class ClioRunTool(IClioRunExecutor executor) {
...
 		CancellationToken cancellationToken = default)
-		=> executor.RunAsync(command, args, destructiveSurface: false, context, cancellationToken);
+		=> executor.RunAsync(command, DictionaryToElement(args), destructiveSurface: false, context, cancellationToken);
+
+	internal static JsonElement? DictionaryToElement(Dictionary<string, JsonElement>? dict)
```
Comment (m-dymytrova) [inline clio/Command/McpServer/Tools/ClioRunTool.cs:475]: **Major — the Dictionary→JsonElement conversion is unit-tested only for the flat non-null and null paths of `clio-run`. Two changed paths stay unguarded.**

1. **Wrapped call shape** `{"args":{"command":"X","args":{...}}}` — with the new `Dictionary<string, JsonElement>?` binding, `args` now binds to a dictionary and relies on this helper re-serializing to an object equivalent to what `RecoverWrappedCall` consumed before. That is exactly repro step 3 in ENG-92653 and the branch most affected by the `JsonElement? → Dictionary<string,JsonElement>?` change, yet no test exercises it. (Correctness looks preserved on inspection, but the SDK by-name-object binding that makes it work is unverified from source — a test would convert that to certainty.)
2. **`clio-run-destructive`** has only a schema-guard test (`...Destructive...InEmittedInputSchema`); its Dictionary→JsonElement conversion (the `ClioRunTool.DictionaryToElement(args)` call on the alias) has no conversion test.

Add a wrapped-form conversion test and a `clio-run-destructive` conversion test (flat + null) alongside the existing two.

**Review by m-dymytrova — CHANGES_REQUESTED** [review]
## PR Review — request changes

Multi-agent review (code-quality, security, performance, testing, bug-risk) against base `spike/mcp-lazy-schema` @ `70f81c9`, head `c4c3709`. The fix itself is correct, minimal, and well-targeted; the blocker is repo-mandated verification coverage.

### Findings (highest → lowest)

**Major — mandatory `clio.mcp.e2e` coverage is missing for the changed tool (repo policy).**
`clio/Command/McpServer/AGENTS.md` (lines 68–72) and the root `AGENTS.md` MCP maintenance policy make this non-negotiable: *"Every new or changed MCP tool must ship with updated `clio.mcp.e2e` coverage… MCP implementation work is incomplete until the real `clio mcp-server` path is exercised end to end."* This PR changes the `args` binding and emitted schema of `clio-run`/`clio-run-destructive` but adds only in-process unit tests. This is the exact blind spot ENG-92653 slipped through in PR #743 — the PR body itself notes the prior unit tests *"bypassed the SDK parameter binding and emitted schema entirely."* The new `McpServerTool.Create(...)` schema guard is a genuine improvement, but it still builds the tool in-process; it does not exercise the real client→server serialize→deserialize→bind round-trip where this bug manifests in Claude Code. The harness already supports this (`clio.mcp.e2e/Support/Mcp/McpServerSession.CallToolAsync(toolName, argsDict, …)` dispatches over the real wire), so the "extend the harness" exemption does not apply, and the test plan's *"Live Claude Code test"* box is still unchecked. (This matches the existing inline review comment on line 471 — consolidated here rather than duplicated.)
*Fix:* add `ClioRunToolE2ETests` dispatching `clio-run` flat-form **and** wrapped-form through the real server and asserting the `args` object reaches the target tool as an object; include a `clio-run-destructive` case.

**Major — untested changed branches: wrapped-form conversion and destructive-alias conversion.** See inline comment on line 475. The Dictionary→JsonElement conversion is unit-tested only for `clio-run` flat + null. The wrapped shape `{"args":{"command":"X","args":{...}}}` (ticket repro step 3, the branch most affected by the type change) and the `clio-run-destructive` conversion path have no conversion test. Bug-risk analysis found the wrapped path *behaves* correctly, but it rests on the SDK binding `Dictionary<string,JsonElement>?` by-name as an object — an assumption best pinned by the E2E/unit test above.

**Minor — cross-class static-helper coupling.** `ClioRunDestructiveTool.Run` reaches into `ClioRunTool.DictionaryToElement`, coupling two peer tools for a serialization utility. Optional: move the helper to a shared internal static so neither peer owns the other's utility. (Cosmetic if both types share the same file.)

**Minor — non-object `args` error surface changed (intended).** `Dictionary<string,JsonElement>?` binds objects only, so a non-object `args` (array/scalar) is now rejected by the SDK as invalid-params before the method runs, instead of reaching the executor's descriptive `ArgumentException`. Non-object `args` was never valid, so this is not a functional regression — but if the executor's descriptive message was valued for agent recoverability, worth a one-line note in the PR description.

### Cleared (no action)
- **Security — neutral.** Destructive gating (tool identity), the self/cross-dispatch guard (`command`), and the redaction backstop (result path) are all untouched; none read `args` content. Re-serializing `args` cannot weaken them.
- **Performance — negligible.** One extra small-object serialize per interactive `clio-run` call, dwarfed by the subsequent command/network execution.
- **Serializer fidelity — verified equivalent.** `DictionaryToElement` uses default options, but the only delta in `CreateMcpSerializerOptions()` is `AllowOutOfOrderMetadataProperties` (a read-only setting); dictionary keys are emitted verbatim (no `DictionaryKeyPolicy`), and `JsonElement` values round-trip. Same JSON reaches the executor.
- **Schema-guard tests** are well-designed and non-vacuous — they use the same serializer options as runtime and would catch a regression to `JsonElement?`.

### Review Summary
Decision: **request changes.** The fix (type `JsonElement? → Dictionary<string,JsonElement>?` so the emitted schema declares `args` as `type:["object","null"]`) is correct, minimal, security-neutral, and performance-neutral, with strong schema-guard unit tests. It is blocked only by the mandatory `clio.mcp.e2e` coverage that this bug class specifically requires (the unit tests cannot reproduce the wire-serialization failure Claude Code hit), plus the untested wrapped-form and destructive-alias conversion branches.

### Requirement Coverage (ENG-92653)
- Flat-form hidden-tool dispatch delivers `args` as a JSON object — **covered** in code + schema-guard unit test; **unclear** end-to-end (no E2E, live-repro box unchecked).
- Wrapped-form dispatch works — **unclear** (preserved on inspection, not tested).
- `args` schema declares `type:object` — **covered** (schema-guard tests).

Confidence limits: review is against a stacked base (`spike/mcp-lazy-schema`); the actual "works from Claude Code" acceptance criterion is inherently a wire-level behavior that no test in this PR exercises.

File: clio/Command/McpServer/Tools/ClioRunTool.cs (hunk 1 above)
Comment (Alexandr-Kravchuk) [reply clio/Command/McpServer/Tools/ClioRunTool.cs:471]: Good call — added `ClioRunToolE2ETests` in `clio.mcp.e2e` (commit d66ca266). It dispatches both a flat-shape `{command, args}` and a wrapped-shape `{args:{command, args}}` `clio-run` call through the real `mcp-server` and asserts the `args` object reaches the dispatched target (`get-guidance`) over the wire — closing the SDK-binding blind spot the unit tests bypass. Both tests pass on net10.0.

Heads-up unrelated to this fix: the `clio.mcp.e2e` project currently does not compile on the base branch (`spike/mcp-lazy-schema`) because `ToolContractGetToolE2ETests.cs` still references a removed harness API (`McpE2ESettings`/`TestConfiguration.Load`/`ArrangeAsync`). I left that base file untouched; my new test compiles and runs cleanly once it is realigned.

File: clio/Command/McpServer/Tools/ClioRunTool.cs (hunk 2 above)
Comment (Alexandr-Kravchuk) [reply clio/Command/McpServer/Tools/ClioRunTool.cs:475]: Added both missing conversion tests in `ClioRunDispatchTests.cs` (commit d66ca266): a wrapped-shape test (`ClioRunTool_ShouldPreserveWrappedShape_WhenDictionaryContainsCommandAndArgs`) that verifies the `Dictionary`→`JsonElement` conversion keeps the `{command, args}` object intact so `ClioRunExecutor.RecoverWrappedCall` can still recover it, plus `clio-run-destructive` flat + null conversion tests. All 39 `ClioRunDispatch` unit tests pass, and a new E2E test now also exercises the wrapped path end-to-end over the real server.

**Review by tetiana-moshon — COMMENTED** [review]
## Creatio Code Review (re-review) — `approve (minor issues)`

Re-review of the delta since my previous `CHANGES_REQUESTED` (range `c4c3709..d66ca26`, commit "test(mcp): add wrapped-shape + destructive conversion units and clio-run E2E coverage").

### Prior Major — RESOLVED
The missing mandatory `clio.mcp.e2e` coverage is now added in `clio.mcp.e2e/ClioRunToolE2ETests.cs`: two tests dispatch `clio-run` over the **real MCP server** — flat shape `{command, args}` and wrapped shape `{args:{command, args}}` — and assert the args object reaches the target (`get-guidance`) by matching the guidance marker in the wire result. Verified against the codebase:
- Fixture pattern (`McpContractFixtureBase.Arrange(...)`, `Session.CallToolAsync(...)`, `CallToolResult.IsError`) matches the established E2E suite (e.g. `AddPackageDependencyToolE2ETests`).
- Target/marker are correct and stable: `page-schema-handlers` + "clio MCP page-schema handlers guide" exist in `PageSchemaHandlersGuidanceResource.cs:20` and are already asserted by `GuidanceGetToolE2ETests.cs:58`.
- Read-only, environment-free target — the test proves args delivery without a live Creatio env, so it runs in the `McpE2E.NoEnvironment` lane.
- `IsError.Should().NotBeTrue()` correctly treats both null and false as success for the `bool?` flag.

The expanded unit tests are also well-formed (AAA, `because`, `[Description]`): wrapped-shape round-trip preservation, and destructive-surface convert + null paths — closing the coverage gaps in the first commit.

### Remaining Minor (non-blocking, unchanged from prior review)
Non-object `args` (array/scalar/stringified JSON) now fails at SDK binding rather than reaching the executor's friendlier "'args' must be a JSON object" error (`ClioRunTool.cs:397-400`). Compliant clients are unaffected (this is the intended behavior); optional to document or accept.

### Requirement coverage (ENG-92653) — all covered
- `args` emits `type` including `object` — covered (SDK-built schema guard, both tools)
- Dictionary→JsonElement conversion (flat, wrapped, null; both surfaces) — covered (unit)
- Real-protocol / real-client validation — **now covered** (E2E, flat + wrapped over the wire)

### Summary
The follow-up commit fully addresses the blocking finding with high-quality E2E coverage that exercises exactly the wire path the original bug hid behind, plus thorough unit coverage. No Blockers and no remaining high-confidence Major findings; the single open item is the optional Minor error-message note. Approving. Confidence: high — small, fully-traced diff reviewed against source. (The MCP E2E project is not yet wired into CI per repo docs, so these tests will run manually/on the E2E lane rather than on every push.)

**Review by tetiana-moshon — APPROVED** [review]

**Review by m-dymytrova — APPROVED** [review]
## Re-review — approve

Both Major findings from the prior review are resolved by `d66ca26` (test-only; the production fix in `ClioRunTool.cs` is unchanged, so the earlier security/performance/bug-risk/quality clears still hold).

**Resolved**
- **`clio.mcp.e2e` coverage (was Major, mandatory per `McpServer/AGENTS.md`).** New `ClioRunToolE2ETests` dispatches `clio-run` through the real MCP server for both the flat shape `{command, args}` and the wrapped shape `{args:{command, args}}`, targeting the environment-free `get-guidance` tool and asserting the args reached the target (marker `clio MCP page-schema handlers guide`, verified present in `PageSchemaHandlersGuidanceResource`). This is the exact over-the-wire path the unit tests bypassed and the class of bug that slipped through PR #743 — now guarded.
- **Untested conversion branches (was Major).** Added unit tests for the wrapped-shape conversion (inner `command`/`args` survive `DictionaryToElement` for `RecoverWrappedCall`) and for the `clio-run-destructive` flat + null conversion paths.

Both new test files follow repo conventions (categories, `MethodName_Should_When` naming, AAA, `because` on every assertion, `[Description]`).

**Remaining (Minor, optional — not blocking)**
- `ClioRunDestructiveTool.Run` reaches into `ClioRunTool.DictionaryToElement`; could move to a shared helper. Cosmetic.
- Non-object `args` is now rejected by SDK binding (invalid-params) rather than the executor's descriptive `ArgumentException`. Intended tradeoff of declaring `type:object`; worth a one-line note in the PR description if the descriptive message was valued.

### Requirement Coverage (ENG-92653)
- Flat-form hidden-tool dispatch delivers `args` as an object — covered (code + schema-guard unit + E2E).
- Wrapped-form dispatch — covered (unit conversion + E2E).
- `args` schema declares `type:object` — covered (schema-guard tests).

Coverage is now sufficient for merge.
